nodes/matrix/apply_and_join.py

```diff
@@ -17,27 +17,90 @@
 # ##### END GPL LICENSE BLOCK #####
 
 import bpy
-from bpy.props import BoolProperty
-from mathutils import Matrix, Vector
+from bpy.props import BoolProperty, EnumProperty, BoolVectorProperty
+from mathutils import Vector
+import numpy as np
+
 from sverchok.node_tree import SverchCustomTreeNode
-from sverchok.data_structure import (Matrix_generate, updateNode) #, fullList_deep_copy)
+from sverchok.data_structure import updateNode
 from sverchok.utils.sv_mesh_utils import mesh_join
+from sverchok.utils.modules.matrix_utils import matrix_apply_np
+
+
+socket_names = ['Vertices', 'Edges', 'Polygons']
+
+def mesh_join_np(verts, edges, pols, out_np):
+
+    lens = [0]
+    for v in verts:
+        lens.append(lens[-1] + v.shape[0])
+
+    v_out = np.concatenate(verts)
+    e_out, p_out = np.array([]), np.array([])
+
+    if len(edges[0]) > 0:
+        e_out = np.concatenate([edg + l for edg, l in zip(edges, lens)])
+
+    if len(pols[0]) > 0 and out_np[2]:
+
+        if pols[0].dtype == object:
+            p_out = [np.array(p) + l  for pol, l in zip(pols, lens) for p in pol]
+
+        else:
+            p_out = np.concatenate([pol + l for pol, l in zip(pols, lens)])
+    else:
+        p_out = [[v + l for v in p]  for pol, l in zip(pols, lens) for p in pol]
+
+    return v_out, e_out, p_out
```

Comment:
It would be nice to know which input arguments are numpy array and which are not. Also `out_np` is unclear. According the argument name I have thought that this probably some array with mesh or empty for keeping merging meshes. This looks very magical - `out_np[2]`.

```suggestion
def mesh_join_np(verts, edges, pols, out_np):

    accum_vert_lens = np.add.accumulate([len(v) for v in chain([[]], verts)])

    is_edges_first_object = len(edges[0]) > 0
    if is_edges_first_object:
        e_out = np.concatenate([edg + l for edg, l in zip(edges, accum_vert_lens)])
        #                      is it really edge??
    else:
        e_out = np.array([], np.int32)

    is_pols_first_object = len(pols[0]) > 0
    is_something_else = out_np[2]
    if is_pols_first_object and is_something_else:
        
        is_array_of_lists = pols[0].dtype == object
        if is_array_of_lists:
            p_out = [np.array(p) + l for pol, l in zip(pols, accum_vert_lens) for p in pol]

        else:
            p_out = np.concatenate([pol + l for pol, l in zip(pols, accum_vert_lens)])
    else:
        p_out = [[v + l for v in p] for pol, l in zip(pols, accum_vert_lens) for p in pol]

    v_out = np.concatenate(verts)
    return v_out, e_out, p_out
```

Reply:
I took part of your suggestions, thanks for the feedback. I hope now everything is more clear

Reply:
Why not using `else` condition?
```Python
    if are_some_edges:
         e_out = np.concatenate([edg + l for edg, l in zip(edges, accum_vert_lens)])
    else:
        e_out = np.array([], np.int32)
```
Reading this code it is clear if there are no edges the functions returns empty array. In your case reader should search this logic somewhere else in the code. 
And I think new type of empty arrays should be set explicitly.


The same with `v_out` variable. To know what is assign to this variable I should look through nearly all function. It is logically to move assigning value to variable as close to useing of the variable as possible.
```Python
    v_out = np.concatenate(verts)
    return v_out, e_out, p_out
```

Reply:
I agree about the else condition.
About the verts line for me feels more natural if is the first input of the node, the first argument of the function, and the first output of the function and the node to be resolved first in the function not in the last line